File: convex-hull-algorithm/convexhull-test.py
# 
# Convex hull algorithm - Test suite (Python)
# 
# Copyright (c) 2021 Project Nayuki
# https://www.nayuki.io/page/convex-hull-algorithm
# 
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Lesser General Public License for more details.
# 
# You should have received a copy of the GNU Lesser General Public License
# along with this program (see COPYING.txt and COPYING.LESSER.txt).
# If not, see <http://www.gnu.org/licenses/>.
# 

import random, unittest
from typing import List, Sequence, Tuple
import convexhull


class ConvexHullTest(unittest.TestCase):
	
	# ---- Fixed test vectors ----
	
	def test_empty(self) -> None:
		points: List[Tuple[float,float]] = []
		actual: List[Tuple[float,float]] = convexhull.make_hull(points)
		expect: List[Tuple[float,float]] = []
		self.assertEqual(expect, actual)
	
	
	def test_one(self) -> None:
		points: List[Tuple[float,float]] = [(3, 1)]
		actual: List[Tuple[float,float]] = convexhull.make_hull(points)
		expect: List[Tuple[float,float]] = points
		self.assertEqual(expect, actual)
	
	
	def test_two_duplicate(self) -> None:
		points: List[Tuple[float,float]] = [(0, 0), (0, 0)]
		actual: List[Tuple[float,float]] = convexhull.make_hull(points)
		expect: List[Tuple[float,float]] = [(0, 0)]
		self.assertEqual(expect, actual)
	
	
	def test_two_horizontal0(self) -> None:
		points: List[Tuple[float,float]] = [(2, 0), (5, 0)]
		actual: List[Tuple[float,float]] = convexhull.make_hull(points)
		expect: List[Tuple[float,float]] = points
		self.assertEqual(expect, actual)
	
	
	def test_two_horizontal1(self) -> None:
		points: List[Tuple[float,float]] = [(-6, -3), (-8, -3)]
		actual: List[Tuple[float,float]] = convexhull.make_hull(points)
		expect: List[Tuple[float,float]] = [(-8, -3), (-6, -3)]
		self.assertEqual(expect, actual)
	
	
	def test_two_vertical0(self) -> None:
		points: List[Tuple[float,float]] = [(1, -4), (1, 4)]
		actual: List[Tuple[float,float]] = convexhull.make_hull(points)
		expect: List[Tuple[float,float]] = points
		self.assertEqual(expect, actual)
	
	
	def test_two_vertical1(self) -> None:
		points: List[Tuple[float,float]] = [(-1, 2), (-1, -3)]
		actual: List[Tuple[float,float]] = convexhull.make_hull(points)
		expect: List[Tuple[float,float]] = [(-1, -3), (-1, 2)]
		self.assertEqual(expect, actual)
	
	
	def test_two_diagonal0(self) -> None:
		points: List[Tuple[float,float]] = [(-2, -3), (2, 0)]
		actual: List[Tuple[float,float]] = convexhull.make_hull(points)
		expect: List[Tuple[float,float]] = points
		self.assertEqual(expect, actual)
	
	
	def test_two_diagonal1(self) -> None:
		points: List[Tuple[float,float]] = [(-2, 3), (2, 0)]
		actual: List[Tuple[float,float]] = convexhull.make_hull(points)
		expect: List[Tuple[float,float]] = points
		self.assertEqual(expect, actual)
	
	
	def test_rectangle(self) -> None:
		points: List[Tuple[float,float]] = [(-3, 2), (1, 2), (1, -4), (-3, -4)]
		actual: List[Tuple[float,float]] = convexhull.make_hull(points)
		expect: List[Tuple[float,float]] = [(-3, -4), (-3, 2), (1, 2), (1, -4)]
		self.assertEqual(expect, actual)
	
	
	
	# ---- Randomized testing ----
	
	def test_horizontal_randomly(self) -> None:
		TRIALS: int = 10000
		for _ in range(TRIALS):
			numpoints: int = random.randrange(30) + 1
			if random.random() < 0.5:
				y: float = random.gauss(0, 1)
				points: List[Tuple[float,float]] = [(random.gauss(0, 1), y) for _ in range(numpoints)]
			else:
				y = random.randrange(20) - 10
				points = [(random.randrange(30), y) for _ in range(numpoints)]
			actual: List[Tuple[float,float]] = convexhull.make_hull(points)
			expected: List[Tuple[float,float]] = [min(points)]
			if max(points) != min(points):
				expected.append(max(points))
			self.assertEqual(actual, expected)
	
	
	def test_vertical_randomly(self) -> None:
		TRIALS: int = 10000
		for _ in range(TRIALS):
			numpoints: int = random.randrange(30) + 1
			if random.random() < 0.5:
				x: float = random.gauss(0, 1)
				points: List[Tuple[float,float]] = [(x, random.gauss(0, 1)) for _ in range(numpoints)]
			else:
				x = random.randrange(20) - 10
				points = [(x, random.randrange(30)) for _ in range(numpoints)]
			actual: List[Tuple[float,float]] = convexhull.make_hull(points)
			expected: List[Tuple[float,float]] = [min(points)]
			if max(points) != min(points):
				expected.append(max(points))
			self.assertEqual(actual, expected)
	
	
	def test_vs_naive_randomly(self) -> None:
		TRIALS: int = 10000
		for _ in range(TRIALS):
			numpoints: int = random.randrange(100)
			if random.random() < 0.5:
				points: List[Tuple[float,float]] = [(random.gauss(0, 1), random.gauss(0, 1)) for _ in range(numpoints)]
			else:
				points = [(random.randrange(10), random.randrange(10)) for _ in range(numpoints)]
			actual: List[Tuple[float,float]] = convexhull.make_hull(points)
			expected: List[Tuple[float,float]] = ConvexHullTest.make_hull_naive(points)
			self.assertEqual(actual, expected)
	
	
	def test_hull_properties_randomly(self) -> None:
		TRIALS: int = 10000
		for _ in range(TRIALS):
			
			# Generate random points
			numpoints: int = random.randrange(100)
			if random.random() < 0.5:
				points: List[Tuple[float,float]] = [(random.gauss(0, 1), random.gauss(0, 1)) for _ in range(numpoints)]
			else:
				points = [(random.randrange(10), random.randrange(10)) for _ in range(numpoints)]
			
			# Compute hull and check properties
			hull: List[Tuple[float,float]] = convexhull.make_hull(points)
			self.assertTrue(ConvexHullTest.is_polygon_convex(hull))
			for p in points:
				self.assertTrue(ConvexHullTest.is_point_in_convex_polygon(hull, p))
			
			# Add duplicate points and check new hull
			if len(points) > 0:
				dupe: int = random.randrange(10) + 1
				for _ in range(dupe):
					points.append(points[random.randrange(len(points))])
				nexthull: List[Tuple[float,float]] = convexhull.make_hull(points)
				self.assertEqual(hull, nexthull)
	
	
	@staticmethod
	def make_hull_naive(points: Sequence[Tuple[float,float]]) -> List[Tuple[float,float]]:
		if len(points) <= 1:
			return list(points)
		
		# Jarvis march / gift wrapping algorithm
		result: List[Tuple[float,float]] = []
		point: Tuple[float,float] = min(points)
		while True:
			result.append(point)
			next: Tuple[float,float] = points[0]
			for p in points:
				ax: float = next[0] - point[0]
				ay: float = next[1] - point[1]
				bx: float = p[0] - point[0]
				by: float = p[1] - point[1]
				cross: float = ax * by - ay * bx
				if cross > 0 or (cross == 0 and bx * bx + by * by > ax * ax + ay * ay):
					next = p
			point = next
			if point == result[0]:
				break
		return result
	
	
	@staticmethod
	def is_polygon_convex(points: Sequence[Tuple[float,float]]) -> bool:
		signum: int = 0
		for i in range(len(points) - 2):
			px, py = points[i + 0]
			qx, qy = points[i + 1]
			rx, ry = points[i + 2]
			sign: int = ConvexHullTest.signum((qx - px) * (ry - qy) - (qy - py) * (rx - qx))
			if sign == 0:
				continue
			elif signum == 0:
				signum = sign
			elif sign != signum:
				return False
		return True
	
	
	@staticmethod
	def is_point_in_convex_polygon(polygon: Sequence[Tuple[float,float]], point: Tuple[float,float]) -> bool:
		signum: int = 0
		for (i, (px, py)) in enumerate(polygon):
			qx, qy = polygon[(i + 1) % len(polygon)]
			sign: int = ConvexHullTest.signum((qx - px) * (point[1] - qy) - (qy - py) * (point[0] - qx))
			if sign == 0:
				continue
			elif signum == 0:
				signum = sign
			elif sign != signum:
				return False
		return True
	
	
	@staticmethod
	def signum(x: float) -> int:
		if x > 0:
			return +1
		elif x < 0:
			return -1
		else:
			return 0



if __name__ == "__main__":
	unittest.main()

File: convex-hull-algorithm/ConvexHullTest.cpp
/* 
 * Convex hull algorithm - Test suite (C++)
 * 
 * Copyright (c) 2022 Project Nayuki
 * https://www.nayuki.io/page/convex-hull-algorithm
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program (see COPYING.txt and COPYING.LESSER.txt).
 * If not, see <http://www.gnu.org/licenses/>.
 */

#include <algorithm>
#include <cstddef>
#include <cstdlib>
#include <exception>
#include <iostream>
#include <random>
#include <utility>
#include "ConvexHull.hpp"

using std::size_t;
using std::vector;


// Forward declarations
static void testEmpty();
static void testOne();
static void testTwoDuplicate();
static void testTwoHorizontal0();
static void testTwoHorizontal1();
static void testTwoVertical0();
static void testTwoVertical1();
static void testTwoDiagonal0();
static void testTwoDiagonal1();
static void testRectangle();
static void testHorizontalRandomly();
static void testVerticalRandomly();
static void testVsNaiveRandomly();
static void testHullPropertiesRandomly();

static vector<Point> makeHullNaive(const vector<Point> &points);
static bool isPolygonConvex(const vector<Point> &points);
static bool isPointInConvexPolygon(const vector<Point> &polygon, const Point &point);
static int signum(double x);


// Random number generation global variables
static std::default_random_engine randGen((std::random_device())());
static std::bernoulli_distribution boolDist;
static std::normal_distribution<double> normalDist;


int main() {
	try {
		testEmpty();
		testOne();
		testTwoDuplicate();
		testTwoHorizontal0();
		testTwoHorizontal1();
		testTwoVertical0();
		testTwoVertical1();
		testTwoDiagonal0();
		testTwoDiagonal1();
		testRectangle();
		testHorizontalRandomly();
		testVerticalRandomly();
		testVsNaiveRandomly();
		testHullPropertiesRandomly();
		
		std::cerr << "Test passed" << std::endl;
		return EXIT_SUCCESS;
	} catch (std::exception &e) {
		std::cerr << e.what() << std::endl;
		return EXIT_FAILURE;
	}
}



/*---- Fixed test vectors ----*/

static void testEmpty() {
	const vector<Point> points{};
	const vector<Point> actual = makeConvexHull(points);
	const vector<Point> expect = points;
	if (actual != expect)
		throw std::runtime_error("Value mismatch");
}


static void testOne() {
	const vector<Point> points{Point{3, 1}};
	const vector<Point> actual = makeConvexHull(points);
	const vector<Point> expect = points;
	if (actual != expect)
		throw std::runtime_error("Value mismatch");
}


static void testTwoDuplicate() {
	const vector<Point> points{Point{0, 0}, Point{0, 0}};
	const vector<Point> actual = makeConvexHull(points);
	const vector<Point> expect{Point{0, 0}};
	if (actual != expect)
		throw std::runtime_error("Value mismatch");
}


static void testTwoHorizontal0() {
	const vector<Point> points{Point{2, 0}, Point{5, 0}};
	const vector<Point> actual = makeConvexHull(points);
	const vector<Point> expect = points;
	if (actual != expect)
		throw std::runtime_error("Value mismatch");
}


static void testTwoHorizontal1() {
	const vector<Point> points{Point{-6, -3}, Point{-8, -3}};
	const vector<Point> actual = makeConvexHull(points);
	const vector<Point> expect{Point{-8, -3}, Point{-6, -3}};
	if (actual != expect)
		throw std::runtime_error("Value mismatch");
}


static void testTwoVertical0() {
	const vector<Point> points{Point{1, -4}, Point{1, 4}};
	const vector<Point> actual = makeConvexHull(points);
	const vector<Point> expect = points;
	if (actual != expect)
		throw std::runtime_error("Value mismatch");
}


static void testTwoVertical1() {
	const vector<Point> points{Point{-1, 2}, Point{-1, -3}};
	const vector<Point> actual = makeConvexHull(points);
	const vector<Point> expect{Point{-1, -3}, Point{-1, 2}};
	if (actual != expect)
		throw std::runtime_error("Value mismatch");
}


static void testTwoDiagonal0() {
	const vector<Point> points{Point{-2, -3}, Point{2, 0}};
	const vector<Point> actual = makeConvexHull(points);
	const vector<Point> expect = points;
	if (actual != expect)
		throw std::runtime_error("Value mismatch");
}


static void testTwoDiagonal1() {
	const vector<Point> points{Point{-2, 3}, Point{2, 0}};
	const vector<Point> actual = makeConvexHull(points);
	const vector<Point> expect = points;
	if (actual != expect)
		throw std::runtime_error("Value mismatch");
}


static void testRectangle() {
	const vector<Point> points{Point{-3, 2}, Point{1, 2}, Point{1, -4}, Point{-3, -4}};
	const vector<Point> actual = makeConvexHull(points);
	const vector<Point> expect{Point{-3, -4}, Point{-3, 2}, Point{1, 2}, Point{1, -4}};
	if (actual != expect)
		throw std::runtime_error("Value mismatch");
}



/*---- Randomized testing ----*/

static void testHorizontalRandomly() {
	const long TRIALS = 100000;
	std::uniform_int_distribution<size_t> numPointsDist(1, 30);
	for (long i = 0; i < TRIALS; i++) {
		size_t len = numPointsDist(randGen);
		vector<Point> points;
		if (boolDist(randGen)) {
			double y = normalDist(randGen);
			for (size_t j = 0; j < len; j++)
				points.push_back(Point{normalDist(randGen), y});
		} else {
			std::uniform_int_distribution<int> valDist(-10, 9);
			double y = valDist(randGen);
			for (size_t j = 0; j < len; j++)
				points.push_back(Point{static_cast<double>(valDist(randGen)), y});
		}
		
		const vector<Point> actual = makeConvexHull(points);
		vector<Point> expected{*std::min_element(points.cbegin(), points.cend())};
		const Point &temp = *std::max_element(points.cbegin(), points.cend());
		if (temp != expected.front())
			expected.push_back(temp);
		if (actual != expected)
			throw std::runtime_error("Value mismatch");
	}
}


static void testVerticalRandomly() {
	const long TRIALS = 100000;
	std::uniform_int_distribution<size_t> numPointsDist(1, 30);
	for (long i = 0; i < TRIALS; i++) {
		size_t len = numPointsDist(randGen);
		vector<Point> points;
		if (boolDist(randGen)) {
			double x = normalDist(randGen);
			for (size_t j = 0; j < len; j++)
				points.push_back(Point{x, normalDist(randGen)});
		} else {
			std::uniform_int_distribution<int> valDist(-10, 9);
			double x = valDist(randGen);
			for (size_t j = 0; j < len; j++)
				points.push_back(Point{x, static_cast<double>(valDist(randGen))});
		}
		
		const vector<Point> actual = makeConvexHull(points);
		vector<Point> expected{*std::min_element(points.cbegin(), points.cend())};
		const Point &temp = *std::max_element(points.cbegin(), points.cend());
		if (temp != expected.front())
			expected.push_back(temp);
		if (actual != expected)
			throw std::runtime_error("Value mismatch");
	}
}


static void testVsNaiveRandomly() {
	const long TRIALS = 100000;
	std::uniform_int_distribution<size_t> numPointsDist(0, 99);
	for (long i = 0; i < TRIALS; i++) {
		size_t len = numPointsDist(randGen);
		vector<Point> points;
		if (boolDist(randGen)) {
			for (size_t j = 0; j < len; j++)
				points.push_back(Point{normalDist(randGen), normalDist(randGen)});
		} else {
			std::uniform_int_distribution<int> valDist(0, 9);
			for (size_t j = 0; j < len; j++)
				points.push_back(Point{static_cast<double>(valDist(randGen)), static_cast<double>(valDist(randGen))});
		}
		
		const vector<Point> expected = makeHullNaive(points);
		const vector<Point> actual = makeConvexHull(std::move(points));
		if (actual != expected)
			throw std::runtime_error("Value mismatch");
	}
}


static void testHullPropertiesRandomly() {
	const long TRIALS = 100000;
	std::uniform_int_distribution<size_t> numPointsDist(0, 99);
	std::uniform_int_distribution<size_t> dupePointsDist(1, 10);
	for (long i = 0; i < TRIALS; i++) {
		
		// Generate random points
		size_t len = numPointsDist(randGen);
		vector<Point> points;
		if (boolDist(randGen)) {
			for (size_t j = 0; j < len; j++)
				points.push_back(Point{normalDist(randGen), normalDist(randGen)});
		} else {
			std::uniform_int_distribution<int> valDist(0, 9);
			for (size_t j = 0; j < len; j++)
				points.push_back(Point{static_cast<double>(valDist(randGen)), static_cast<double>(valDist(randGen))});
		}
		
		// Compute hull and check properties
		const vector<Point> hull = makeConvexHull(points);
		if (!isPolygonConvex(hull))
			throw std::runtime_error("Polygon not convex");
		for (const Point &p : points) {
			if (!isPointInConvexPolygon(hull, p))
				throw std::runtime_error("Point not in polygon");
		}
		
		// Add duplicate points and check new hull
		if (!points.empty()) {
			size_t dupe = dupePointsDist(randGen);
			for (size_t j = 0; j < dupe; j++) {
				std::uniform_int_distribution<size_t> sizeDist(0, points.size() - 1);
				points.push_back(points.at(sizeDist(randGen)));
			}
			const vector<Point> nextHull = makeConvexHull(std::move(points));
			if (nextHull != hull)
				throw std::runtime_error("Value mismatch");
		}
	}
}


static vector<Point> makeHullNaive(const vector<Point> &points) {
	if (points.size() <= 1)
		return vector<Point>(points);
	
	// Jarvis march / gift wrapping algorithm
	vector<Point> result;
	const Point *point = &*std::min_element(points.cbegin(), points.cend());
	do {
		result.push_back(*point);
		const Point *next = &points.front();
		for (const Point &p : points) {
			double ax = next->x - point->x;
			double ay = next->y - point->y;
			double bx = p.x - point->x;
			double by = p.y - point->y;
			double cross = ax * by - ay * bx;
			if (cross > 0 || (cross == 0 && bx * bx + by * by > ax * ax + ay * ay))
				next = &p;
		}
		point = next;
	} while (*point != result.front());
	return result;
}


static bool isPolygonConvex(const vector<Point> &points) {
	int state = 0;
	for (size_t i = 0; i + 2 < points.size(); i++) {
		const Point &p = points.at(i + 0);
		const Point &q = points.at(i + 1);
		const Point &r = points.at(i + 2);
		int sign = signum((q.x - p.x) * (r.y - q.y) - (q.y - p.y) * (r.x - q.x));
		if (sign == 0)
			continue;
		else if (state == 0)
			state = sign;
		else if (sign != state)
			return false;
	}
	return true;
}


static bool isPointInConvexPolygon(const vector<Point> &polygon, const Point &point) {
	int state = 0;
	for (size_t i = 0; i < polygon.size(); i++) {
		const Point &p = polygon.at(i);
		const Point &q = polygon.at((i + 1) % polygon.size());
		int sign = signum((q.x - p.x) * (point.y - q.y) - (q.y - p.y) * (point.x - q.x));
		if (sign == 0)
			continue;
		else if (state == 0)
			state = sign;
		else if (sign != state)
			return false;
	}
	return true;
}


static int signum(double x) {
	if (x > 0)
		return +1;
	else if (x < 0)
		return -1;
	else
		return 0;
}
